### PR TITLE
UV Event Loop (PECL ext-uv)

### PR DESCRIPTION
**Pull request by WyriHaximus**

### CODE REVIEW

File: src/Factory.php [hunk 1]
```diff
@@ -7,7 +7,9 @@ class Factory
     public static function create()
     {
         // @codeCoverageIgnoreStart
-        if (class_exists('libev\EventLoop', false)) {
+        if (function_exists('uv_default_loop')) {
+            return new LibUvLoop();
```
Comment (kelunik) [inline src/Factory.php]: Drop the `()` for consistency.

File: src/LibUvLoop.php [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+        $this->running = true;
+
+        while ($this->running) {
+            $this->futureTickQueue->tick();
+
+            if ($this->futureTickQueue->isEmpty() && empty($this->events) && $this->timerEvents->count() === 0) {
+                break;
+            }
+
+            \uv_run($this->uv, \UV::RUN_NOWAIT);
```
Comment (kelunik) [inline src/LibUvLoop.php]: Always using `RUN_NOWAIT` will result in a busy loop.

File: src/LibUvLoop.php [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+            return;
+        }
+
+        unset($this->listeners[(int) $stream]['read']);
+
+        if (!isset($this->listeners[(int) $stream]['read'])
+            && !isset($this->listeners[(int) $stream]['write'])) {
+            unset($this->events[(int) $stream]);
+        }
+    }
```
Comment (clue) [inline src/LibUvLoop.php]: Isn't this supposed to call `uv_poll_stop()` if this stream does no longer have a listener? (see also below)

File: src/Factory.php (hunk 1 above)
Comment (clue) [reply src/Factory.php]: I'd rather keep this as-is and eventually address the other two as CS violations :+1:

File: travis-init.sh [hunk 4]
```diff
@@ -34,4 +34,17 @@ if [[ "$TRAVIS_PHP_VERSION" != "hhvm" &&
         echo "extension=libev.so" >> "$(php -r 'echo php_ini_loaded_file();')"
     fi
 
+    # install 'libev' PHP extension (does not support php 7)
```
Comment (clue) [inline travis-init.sh]: Copy pasta? :spaghetti:

Comment (WyriHaximus) [reply travis-init.sh]: Yeah I went for a 🍕 but ended up with a 🍝 instead,

File: src/LibUvLoop.php (hunk 2 above)
Comment (WyriHaximus) [reply src/LibUvLoop.php]: What do you suggest as an alternative?

File: src/Factory.php (hunk 1 above)
Comment (WyriHaximus) [reply src/Factory.php]: Lets fix this in a follow up PR as it's 2 vs. 2 right now for both styles

File: src/LibUvLoop.php (hunk 3 above)
Comment (WyriHaximus) [reply src/LibUvLoop.php]: Yup, just pushed changes that address that 👍

File: tests/AbstractLoopTest.php [hunk 5]
```diff
@@ -428,9 +428,8 @@ public function testSignalsKeepTheLoopRunning()
     {
         $function = function () {};
         $this->loop->addSignal(SIGUSR1, $function);
-        $this->loop->addTimer(1.5, function () use ($function) {
+        $this->loop->addTimer(1.6, function () use ($function) {
```
Comment (docteurklein) [inline tests/AbstractLoopTest.php]: What is the reason for this change?

File: src/LibUvLoop.php (hunk 2 above)
Comment (kelunik) [reply src/LibUvLoop.php]: Depends on whether you have things scheduled for the next tick or not, see https://github.com/amphp/amp/blob/1cb074ac9078f1c89787b539d605083c5845ceab/lib/Loop/UvDriver.php#L174

File: src/Factory.php [hunk 6]
```diff
@@ -24,13 +24,15 @@
     public static function create()
     {
         // @codeCoverageIgnoreStart
-        if (function_exists('uv_default_loop')) {
+        $isPHPSeven = PHP_VERSION_ID >= 70000;
```
Comment (kelunik) [inline src/Factory.php]: Why that? `PHP_VERSION_ID >= 70000` can be optimized out by OPcache, but sure if this can. You can also use `PHP_MAJOR_VERSION >= 7`.

File: src/LibUvLoop.php (hunk 2 above)
Comment (WyriHaximus) [reply src/LibUvLoop.php]: Addressed this, thanks :+1:

**Comment by WyriHaximus** [comment]
@clue @jsor right, rebased this PR against the latest `master`. Could you review?

File: src/LibUvLoop.php [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+namespace React\EventLoop;
+
+use React\EventLoop\Tick\FutureTickQueue;
+use React\EventLoop\Timer\Timer;
+use SplObjectStorage;
+
+/**
+ * @see https://github.com/bwoebi/php-uv
+ */
+final class LibUvLoop implements LoopInterface
```
[the remark this answers is not in the document]
Comment (kelunik) [reply src/LibUvLoop.php]: Rather `ExtUvLoop` then.

[the remark this answers is not in the document]
Comment (clue) [reply src/LibUvLoop.php]: @kelunik What name does this extension use in `php -m`? This PR currently `suggest`s installing `ext-libuv` :v:

[the remark this answers is not in the document]
Comment (kelunik) [reply src/LibUvLoop.php]: `uv`

[the remark this answers is not in the document]
Comment (clue) [reply src/LibUvLoop.php]: @kelunik Thank you for confirming! In this case the `suggest` keyword should be updated and `ExtUvLoop` would indeed be the correct name :+1:

File: tests/AbstractLoopTest.php [hunk 8]
```diff
@@ -557,9 +557,8 @@ public function testSignalsKeepTheLoopRunning()
         $loop = $this->loop;
         $function = function () {};
         $this->loop->addSignal(SIGUSR1, $function);
-        $this->loop->addTimer(1.5, function () use ($function, $loop) {
+        $this->loop->addTimer(1.6, function () use ($function, $loop) {
```
[the remark this answers is not in the document]
Comment (WyriHaximus) [reply tests/AbstractLoopTest.php]: Something broke (it's been a while :zipper_mouth_face: )

File: src/LibUvLoop.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (WyriHaximus) [reply src/LibUvLoop.php]: Consider it done :+1:

File: src/LibUvLoop.php [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+    private function addStream($stream)
+    {
+        if (!isset($this->streamEvents[(int) $stream])) {
+            $this->streamEvents[(int) $stream] = \uv_poll_init_socket($this->uv, $stream);
+        }
+
+        // Run in tick or else things epically fail with loop->watchers[w->fd] == w
+        $this->futureTick(function () use ($stream) {
+            $this->pollStream($stream);
+        });
```
[the remark this answers is not in the document]
Comment (WyriHaximus) [reply src/LibUvLoop.php]: Comment this tick wrapping code out. Either it fails locally or on Travis and on test suite or using examples

File: src/Factory.php [hunk 10]
```diff
@@ -24,14 +24,17 @@
     public static function create()
     {
         // @codeCoverageIgnoreStart
-        if (class_exists('libev\EventLoop', false)) {
+        if (function_exists('uv_default_loop') && PHP_MAJOR_VERSION === 7) {
```
[the remark this answers is not in the document]
Comment (WyriHaximus) [reply src/Factory.php]: Done :+1:

File: src/LibUvLoop.php [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+        $callback = function () use ($timer) {
+            call_user_func($timer->getCallback(), $timer);
+        };
+
+        $event = \uv_timer_init($this->uv);
+        $this->timerEvents->attach($timer, $event);
+        \uv_timer_start(
+            $event,
+            $interval * 1000,
+            $interval * 1000,
```
[the remark this answers is not in the document]
Comment (WyriHaximus) [reply src/LibUvLoop.php]: The cast to integer is required because on float it errors with: `uv_timer_start() expects parameter 2 to be integer, float given`

Also added `ceil()`, the difference with a only a cast is that the cast behaves like `floor()`. So with `ceil()` we enforce the timer runs after it's intended run time where `floor()` might run it just before. But this is at the milisecond level so it won't matter that much

File: tests/AbstractLoopTest.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (WyriHaximus) [reply tests/AbstractLoopTest.php]: Can't find any errors locally after reverting this so absolutely no clue anymore why I changed this :zipper_mouth_face:

**Comment by andreybolonin** [comment]
what about to merge on this pr?
@WyriHaximus @clue @cboden

File: src/ExtUvLoop.php [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+
+namespace React\EventLoop;
+
+use React\EventLoop\Tick\FutureTickQueue;
+use React\EventLoop\Timer\Timer;
+use SplObjectStorage;
+
+/**
+ * @see https://github.com/bwoebi/php-uv
+ */
```
Comment (clue) [inline src/ExtUvLoop.php]: This package is (now) available via PECL (https://pecl.php.net/package/uv), so this PR should probably be updated to use reference official `ext-uv` instead of installing from source :+1: Docblock and README should be in line with `ext-ev`.

File: src/ExtUvLoop.php [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+
+    private function addStream($stream)
+    {
+        if (!isset($this->streamEvents[(int) $stream])) {
+            $this->streamEvents[(int) $stream] = \uv_poll_init_socket($this->uv, $stream);
+        }
+
+        // Run in tick or else things epically fail with loop->watchers[w->fd] == w
+        $this->futureTick(function () use ($stream) {
+            $this->pollStream($stream);
```
Comment (clue) [inline src/ExtUvLoop.php]: I'd like to have a better understanding of why this is needed and what issue this addresses. Is this an upstream issue? Could this possibly be related to https://github.com/bwoebi/php-uv/issues/49?

It's my understanding that we have a single `uv_poll_init_socket()` for a single stream with both a read and write side and may call `uv_poll_start()` multiple times on this same handle and will call `uv_poll_stop()` once both sides have been removed. This is covered by the `libuv` documentation as well as `ext-uv`, so I'm failing to see a problem here and how deferring this to the next tick could address any issues?

File: travis-init.sh [hunk 14]
```diff
@@ -39,4 +39,18 @@ if [[ "$TRAVIS_PHP_VERSION" != "hhvm" &&
...
+          "$TRAVIS_PHP_VERSION" = "7.2" ]]; then
+        git clone --recursive https://github.com/bwoebi/php-uv
+        pushd php-uv
+        phpize
+        ./configure
+        make
+        make install
+        popd
+        echo "extension=uv.so" >> "$(php -r 'echo php_ini_loaded_file();')"
+    fi
```
Comment (clue) [inline travis-init.sh]: Should this install from PECL instead (see above)?

File: tests/ExtUvLoopTest.php [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+
+class ExtUvLoopTest extends AbstractLoopTest
+{
+    public function createLoop()
+    {
+        if (!function_exists('uv_loop_new')) {
+            $this->markTestSkipped('uv tests skipped because ext-uv is not installed.');
+        }
+
+        return new ExtUvLoop();
```
Comment (clue) [inline tests/ExtUvLoopTest.php]: Similar logic should be implemented for timer tests?

Comment (WyriHaximus) [reply tests/ExtUvLoopTest.php]: Yup done :+1:

File: travis-init.sh (hunk 14 above)
Comment (WyriHaximus) [reply travis-init.sh]: Installing through PECL now :tada:

File: src/ExtUvLoop.php (hunk 12 above)
Comment (WyriHaximus) [reply src/ExtUvLoop.php]: Completely missed that release on PECL, updated the docblock and README :+1:

File: src/ExtUvLoop.php (hunk 13 above)
Comment (WyriHaximus) [reply src/ExtUvLoop.php]: Hmm I'll have a look at doing it that way. This way turned out to be the working solution after hours of debugging. Might have missed that approach 👍

Comment (WyriHaximus) [reply src/ExtUvLoop.php]: With future tick: https://travis-ci.org/reactphp/event-loop/jobs/427051253#L1898
Without future tick: https://travis-ci.org/reactphp/event-loop/jobs/427051579#L1292

Going to dive deeper into the why and see if there is an alternative solution but with a future tick it works.

**Comment by WyriHaximus** [comment]
@clue @jsor @cboden updated PR and it fully works without odd hacks now, could you review this PR again?

File: src/ExtUvLoop.php [hunk 16]
```diff
@@ -0,0 +1,306 @@
...
+
+        if (isset($this->writeStreams[(int) $stream])) {
+            $flags |= \UV::WRITABLE;
+        }
+
+        if (isset($this->flags[(int) $stream]) && $this->flags[(int) $stream] == $flags) {
+            return;
+        }
+
+        $this->flags[(int) $stream] = $flags;
```
Comment (clue) [inline src/ExtUvLoop.php]: Seems like this can be simplified by removing this? Why does this store the flags when they are only to be assigned to the `uv_poll_start()` function? It's my understanding this method will only ever be called when a stream listener changes.

File: src/ExtUvLoop.php [hunk 17]
```diff
@@ -0,0 +1,306 @@
...
+     *
+     * @return callable Returns a callback
+     */
+    private function createStreamListener()
+    {
+        $callback = function ($event, $status, $events, $stream) {
+            if ($status !== 0) {
+                unset($this->flags[(int) $stream]);
+                $this->pollStream($stream);
+            }
```
Comment (clue) [inline src/ExtUvLoop.php]: Do we have a test that covers this block of code? As per http://docs.libuv.org/en/v1.x/poll.html#c.uv_poll_start, it's my understanding that this would indicate an error and not a situation where the poll flags need to be updated?

File: src/Factory.php [hunk 18]
```diff
@@ -24,14 +24,17 @@
     public static function create()
     {
         // @codeCoverageIgnoreStart
-        if (\class_exists('libev\EventLoop', false)) {
+        if (\function_exists('uv_loop_new') && \PHP_MAJOR_VERSION === 7) {
+            // only use ext-uv on PHP 7
+            return new ExtUvLoop();
```
Comment (clue) [inline src/Factory.php]: Is the version check needed here? I would vote to remove this, the extension is only available on PHP 7+ anyway, so this is not a decision we need to make.

**Comment by CharlotteDunois** [comment]
You probably want to add a method to this loop implementation class to allow the user to get the underlying UV event loop, otherwise making use of the UV extension is limited to just the event loop implementation. A libuv-based adapter for react/filesystem will be impossible, as you can't get the UV event loop that is used by this loop implementation. And I don't see that this loop implementation will do 100% coverage of libuv. 😅

**Comment by WyriHaximus** [comment]
@CharlotteDunois there you go https://github.com/reactphp/event-loop/pull/112/commits/5d40fc519721f8af581b69c7d85d2c748839032c this was always my intention to do but I wanted to focus on a stable loop first. But since that cat is out of the bag I've added it. My intention is to add a ext-uv filesystem adapter as well once this PR has been working to my satisfaction. (Currently working out an issue with child processes in a real world project.)

**Comment by WyriHaximus** [comment]
@clue && @jsor && @cboden could you review this PR again. I feel it's ready to be merged and released into the wild :tada: .

**Comment by WyriHaximus** [comment]
FYI the issues I was having wheren't related to the event loop directly but due to file descriptors being passed on child processed by default, solving that down/upstream depending on your perspective: https://github.com/WyriHaximus/reactphp-child-process-messenger/pull/22

File: .travis.yml [hunk 19]
```diff
@@ -32,9 +32,15 @@ cache:
...
+  - sudo apt-get update -q
+  - sudo apt-get install libuv1-dev || true
+
 install:
   - ./travis-init.sh
   - composer install
 
 script:
   - ./vendor/bin/phpunit --coverage-text
+  - php examples/14-http-client-async.php
```
Comment (clue) [inline .travis.yml]: Leftover?

File: src/ExtUvLoop.php [hunk 20]
```diff
@@ -0,0 +1,320 @@
...
+    private $readStreams = array();
+    private $writeStreams = array();
+    private $running;
+    private $signals;
+    private $signalEvents = array();
+    private $streamListener;
+
+    public function __construct()
+    {
+        if (!function_exists('uv_loop_new')) {
```
Comment (clue) [inline src/ExtUvLoop.php]: Should be updated to use global references as per https://github.com/reactphp/event-loop/pull/167 that was merged in the meantime.

File: src/ExtUvLoop.php [hunk 21]
```diff
@@ -0,0 +1,320 @@
...
+        unset($this->writeStreams[(int) $stream]);
+        $this->removeStream($stream);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function addTimer($interval, $callback)
+    {
+        $timer = new Timer( $interval, $callback, false);
```
Comment (clue) [inline src/ExtUvLoop.php]: nit: superfluous space

File: src/ExtUvLoop.php (hunk 17 above)
Comment (clue) [reply src/ExtUvLoop.php]: Ping @WyriHaximus, what's the status here?

File: .travis.yml (hunk 19 above)
Comment (WyriHaximus) [reply .travis.yml]: Yup, fixed :+1:

File: src/ExtUvLoop.php (hunk 20 above)
Comment (WyriHaximus) [reply src/ExtUvLoop.php]: Fixed :+1:

File: src/ExtUvLoop.php (hunk 21 above)
Comment (WyriHaximus) [reply src/ExtUvLoop.php]: Fixed :+1:

File: src/ExtUvLoop.php (hunk 17 above)
Comment (WyriHaximus) [reply src/ExtUvLoop.php]: No status change

Comment (WyriHaximus) [reply src/ExtUvLoop.php]: Really unsure what to do with this so I've removed it. If we run into this we can add it again once we have a reproduceable case.

**Comment by WyriHaximus** [comment]
@clue will squash into one commit once everyone approves it before merging it

File: src/ExtUvLoop.php [hunk 22]
```diff
@@ -0,0 +1,320 @@
...
+
+    public function __construct()
+    {
+        if (!\function_exists('uv_loop_new')) {
+            throw new \BadMethodCallException('Cannot create LibUvLoop, ext-uv extension missing');
+        }
+
+        $this->uv = \uv_loop_new();
+        $this->futureTickQueue = new FutureTickQueue();
+        $this->timers = new SplObjectStorage();
```
Comment (CharlotteDunois) [inline src/ExtUvLoop.php]: Missing \ prefix

File: src/ExtUvLoop.php [hunk 23]
```diff
@@ -0,0 +1,320 @@
...
+        $callback = function () use ($timer) {
+            \call_user_func($timer->getCallback(), $timer);
+        };
+
+        $event = \uv_timer_init($this->uv);
+        $this->timers->attach($timer, $event);
+        \uv_timer_start(
+            $event,
+            (int)($interval * 1000) + 1,
+            (int)($interval * 1000) + 1,
```
Comment (CharlotteDunois) [inline src/ExtUvLoop.php]: Shouldn't these two lines have a space between the type cast and the operation, just like all the other type casts here?

File: src/ExtUvLoop.php [hunk 24]
```diff
@@ -0,0 +1,320 @@
...
+            if ($timers->contains($timer)) {
+                $that->cancelTimer($timer);
+            }
+        };
+
+        $event = \uv_timer_init($this->uv);
+        $this->timers->attach($timer, $event);
+        \uv_timer_start(
+            $event,
+            (int)($interval * 1000) + 1,
```
Comment (CharlotteDunois) [inline src/ExtUvLoop.php]: Shouldn't these two lines have a space between the type cast and the operation, just like all the other type casts here?

File: src/ExtUvLoop.php (hunk 23 above)
Comment (WyriHaximus) [reply src/ExtUvLoop.php]: Yup, updated it

File: src/ExtUvLoop.php (hunk 24 above)
Comment (WyriHaximus) [reply src/ExtUvLoop.php]: Yup, updated it

File: src/ExtUvLoop.php (hunk 17 above)
Comment (WyriHaximus) [reply src/ExtUvLoop.php]: Ok done

**Comment by WyriHaximus** [comment]
It is time for this PR to be merged, I've squashed all commits into one, the original unrewritten history can be found here: https://github.com/WyriHaximus-labs/event-loop/tree/blacklight-history

**Comment by WyriHaximus** [comment]
webp net-gifmaker 1
